src/cameraGeom/TransformMap.cc

```diff
@@ -0,0 +1,171 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include <exception>
+#include <memory>
+#include <sstream>
+#include <type_traits>
+
+#include "lsst/pex/exceptions.h"
+#include "lsst/afw/cameraGeom/TransformMap.h"
+
+namespace lsst {
+namespace afw {
+namespace cameraGeom {
+
+namespace {
+
+/**
+ * Represent a set of camera Transforms as a FrameSet.
+ *
+ * @tparam Map Any type satisfying the STL map API and mapping CameraSys to
+ *             shared_ptr<Transform>.
+ *
+ * @param root Common coordinate system serving as root of FrameSet's tree.
+ * @param transforms A map whose keys are coordinate systems, and whose
+ *                   values point to Transforms that convert from `root` to the
+ *                   corresponding key. All Transforms must be invertible.
+ * @return a newly allocated FrameSet corresponding to the method arguments
+ *
+ * @throws lsst::pex::exceptions::InvalidParameterError Thrown if
+ *         `transforms` contains `root`, or if any Transform is not
+ *         invertible.
+ */
+template <class Map>
+std::unique_ptr<ast::FrameSet> makeTransforms(CameraSys const &root, Map const &transforms) {
+    ast::Frame rootFrame(2, "Ident=" + root.getSysName());
+    auto result = std::unique_ptr<ast::FrameSet>(new ast::FrameSet(rootFrame));
+
+    for (auto keyValue : transforms) {
+        CameraSys const key = keyValue.first;
+        std::shared_ptr<TransformMap::Transform> const value = keyValue.second;
+
+        if (key == root) {
+            std::ostringstream buffer;
+            buffer << "Cannot specify a Transform from " << root << " to itself.";
+            throw LSST_EXCEPT(pex::exceptions::InvalidParameterError, buffer.str());
+        }
+        if (!value->hasForward()) {
+            std::ostringstream buffer;
+            buffer << *value << " from " << key << " has no forward transform.";
+            throw LSST_EXCEPT(pex::exceptions::InvalidParameterError, buffer.str());
+        }
+        if (!value->hasInverse()) {
+            std::ostringstream buffer;
+            buffer << *value << " from " << key << " has no inverse transform.";
+            throw LSST_EXCEPT(pex::exceptions::InvalidParameterError, buffer.str());
+        }
+
+        auto toFrame = value->getFrameSet()->getFrame(ast::FrameSet::CURRENT);
+        toFrame->setIdent(key.getSysName());
+        result->addFrame(ast::FrameSet::BASE, *(value->getFrameSet()->getMapping()), *toFrame);
+    }
+    return result;
+}
+
+/**
+ * Identify the frames in to a FrameSet constructed by makeTransforms.
```

Comment:
I think "to" should be removed? I don't actually remember what I meant to say. :fearful: